From 690a3eec78fa36bec207845ab6787e802d2cdab3 Mon Sep 17 00:00:00 2001
From: Jesse Taube <jesse@rivosinc.com>
Date: Fri, 6 Sep 2024 20:32:13 -0400
Subject: [PATCH] timer

---
 drivers/clocksource/timer-clint.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/clocksource/timer-clint.c b/drivers/clocksource/timer-clint.c
index 0bdd9d7ec545..aab46f2ce27f 100644
--- a/drivers/clocksource/timer-clint.c
+++ b/drivers/clocksource/timer-clint.c
@@ -28,9 +28,9 @@
 #include <asm/clint.h>
 #endif
 
-#define CLINT_IPI_OFF		0
-#define CLINT_TIMER_CMP_OFF	0x4000
-#define CLINT_TIMER_VAL_OFF	0xbff8
+#define CLINT_IPI_OFF		0x1a0
+#define CLINT_TIMER_CMP_OFF	0x1b8
+#define CLINT_TIMER_VAL_OFF	0x1b0
 
 /* CLINT manages IPI and Timer for RISC-V M-mode  */
 static u32 __iomem *clint_ipi_base;
-- 
2.45.2

